File: ios/Runner/TimingBridge.swift
import AVFoundation
import Flutter
import UIKit

@MainActor
final class TimingBridge {
  private let channel: FlutterMethodChannel
  private var beat: AVAudioPlayer?
  private var cue: AVAudioPlayer?
  private var tempo: Int?
  /// Built once and replayed. Creating an AVAudioPlayer per cue put tens of
  /// milliseconds between the moment a phase ends and the sound.
  private var beats: [Int: AVAudioPlayer] = [:]
  private var cues: [String: AVAudioPlayer] = [:]
  private var sessionActive = false
  private var observers: [NSObjectProtocol] = []
  private let speech = AVSpeechSynthesizer()
  private var pendingSpeech: Task<Void, Never>?

  init(messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(name: "setpad/timing", binaryMessenger: messenger)
    channel.setMethodCallHandler { [weak self] call, result in
      guard let self else { result(FlutterMethodNotImplemented); return }
      let args = call.arguments as? [String: Any] ?? [:]
      if call.method == "speak" {
        self.speak(args["text"] as? String ?? "", locale: args["locale"] as? String ?? "en",
                   rate: args["rate"] as? Double ?? 1.15)
        result(nil)
        return
      }
      guard call.method == "configure" else { result(FlutterMethodNotImplemented); return }
      do {
        try self.configure(active: args["active"] as? Bool ?? false,
                           bpm: args["bpm"] as? Int, cueName: args["cue"] as? String,
                           phase: args["phase"] as? Double ?? 0)
        result(nil)
      } catch {
        self.stop()
        result(FlutterError(code: "audioUnavailable", message: nil, details: nil))
      }
    }
    for name in [UIApplication.willResignActiveNotification, AVAudioSession.interruptionNotification] {
      observers.append(NotificationCenter.default.addObserver(forName: name, object: nil, queue: .main) { [weak self] _ in
        Task { @MainActor in self?.stop(); self?.channel.invokeMethod("interrupted", arguments: nil) }
      })
    }
  }

  /// Match Tempo: click, then count half a beat later (at most one second).
  ///
  /// The count comes from Dart's clock and the click from this player, which
  /// started a few tens of milliseconds later. So when a count arrives the click
  /// it belongs to may not have sounded yet (the loop is near its end): then
  /// wait for that click, not for the one after half a beat ago.
  /// A newer count replaces one still waiting. If the previous word is still
  /// sounding, wait a little for it instead of dropping this count or cutting
  /// that one off.
  private func speak(_ text: String, locale: String, rate: Double) {
    guard !text.isEmpty else { return }
    pendingSpeech?.cancel(); pendingSpeech = nil
    var countRemaining = 0.0, patience = 0.15
    if let tempo, let beat {
      let period = 60.0 / Double(tempo), half = min(period / 2, 1.0), at = beat.currentTime
      countRemaining = at > period * 0.75 ? period - at + half : max(0, half - at)
      // Start no later than a quarter beat after the planned moment, so a late
      // word never lands on the next click and lateness cannot pile up.
      patience = period * 0.25
    }
    let cueRemaining = cue.flatMap { $0.isPlaying ? max(0, $0.duration - $0.currentTime) : nil } ?? 0
    let delay = max(countRemaining, cueRemaining)
    pendingSpeech = Task { @MainActor [weak self] in
      if delay > 0 {
        try? await Task.sleep(nanoseconds: UInt64((delay + 0.005) * 1_000_000_000))
      }
      let giveUp = Date().addingTimeInterval(patience)
      while let self, self.speech.isSpeaking, Date() < giveUp, !Task.isCancelled {
        try? await Task.sleep(nanoseconds: 20_000_000)
      }
      guard !Task.isCancelled, let self else { return }
      self.pendingSpeech = nil
      guard !self.speech.isSpeaking else { return }
      let utterance = AVSpeechUtterance(string: text)
      utterance.voice = AVSpeechSynthesisVoice(language: locale)
        ?? AVSpeechSynthesisVoice(language: Locale.current.identifier)
      utterance.rate = min(AVSpeechUtteranceMaximumSpeechRate,
                           AVSpeechUtteranceDefaultSpeechRate * Float(rate))
      self.speech.speak(utterance)
    }
  }

  private func cancelSpeech() {
    pendingSpeech?.cancel(); pendingSpeech = nil
    speech.stopSpeaking(at: .immediate)
  }

  private func configure(active: Bool, bpm: Int?, cueName: String?, phase: Double) throws {
    let session = AVAudioSession.sharedInstance()
    // Activating the session costs tens of milliseconds, so do it once and
    // keep it. Doing it per cue was what put a lag before every beep.
    if (active || cueName != nil), !sessionActive {
      try session.setCategory(.playback, mode: .default, options: [.mixWithOthers])
      try session.setActive(true)
      sessionActive = true
    }
    UIApplication.shared.isIdleTimerDisabled = active
    // The app allows 10 BPM; anything slower than this range is a typo.
    let validTempo = active ? bpm.flatMap { (10...300).contains($0) ? $0 : nil } : nil
    if validTempo != tempo || !active {
      // Stopping cuts the voice. A tempo change while running (Tabata work to
      // rest) only drops a count still waiting: the last number of the round
      // is allowed to finish.
      if active { pendingSpeech?.cancel(); pendingSpeech = nil } else { cancelSpeech() }
      beat?.stop(); beat = nil; tempo = validTempo
      if let bpm = validTempo {
        let player = try beatPlayer(bpm)
        // Start where the timer is inside its beat, not at a fresh click: after a
        // pause or a re-sync the clicks stay on the timer's beat grid.
        player.currentTime = min(max(0, phase), max(0, player.duration - 0.001))
        guard player.play() else { throw NSError(domain: "setpad.timing", code: 1) }
        beat = player
      }
    }
    if let cueName {
      let player = try cuePlayer(cueName)
      if cue !== player { cue?.stop() }
      player.currentTime = 0
      guard player.play() else { throw NSError(domain: "setpad.timing", code: 2) }
      cue = player
      // Build the rest while this one is already sounding, so the first cue of
      // each kind is no slower than the ones after it.
      for name in ["ready", "work", "rest", "complete"] where cues[name] == nil {
        _ = try? cuePlayer(name)
      }
    } else if !active {
      cue?.stop(); cue = nil
      try? session.setActive(false, options: [.notifyOthersOnDeactivation])
      sessionActive = false
    }
  }

  private func beatPlayer(_ bpm: Int) throws -> AVAudioPlayer {
    if let player = beats[bpm] { return player }
    let player = try AVAudioPlayer(data: wave(Self.click, length: 60.0 / Double(bpm)))
    player.numberOfLoops = -1; player.prepareToPlay(); beats[bpm] = player
    return player
  }

  private func cuePlayer(_ name: String) throws -> AVAudioPlayer {
    if let player = cues[name] { return player }
    let player = try AVAudioPlayer(data: wave(Self.cue(name)))
    player.prepareToPlay(); cues[name] = player
    return player
  }

  private func stop() {
    cancelSpeech()
    beat?.stop(); beat = nil; cue?.stop(); cue = nil; tempo = nil
    UIApplication.shared.isIdleTimerDisabled = false
    try? AVAudioSession.sharedInstance().setActive(false, options: [.notifyOthersOnDeactivation])
    sessionActive = false
  }

  /// Pitch, partials, length and loudness measured from Tempo's cue recordings
  /// (bpm.mp3, prebpm.mp3, end_3s.mp3), so the two apps sound like one.
  private struct Tone { let partials: [(Double, Double)]; let seconds: Double; let decay: Bool; let gain: Double }
  private static let click = Tone(partials: [(655, 1), (1965, 0.1)], seconds: 0.12, decay: true, gain: 0.45)
  private static let ding: [(Double, Double)] = [(1787, 1), (2664, 0.38), (1010, 0.29)]
  private static func cue(_ name: String) -> Tone {
    switch name {
    case "ready": return Tone(partials: [(523, 1), (1568, 0.25)], seconds: 0.08, decay: false, gain: 0.5)
    case "work": return Tone(partials: [(1046, 1), (3138, 0.18)], seconds: 0.22, decay: false, gain: 0.5)
    case "complete": return Tone(partials: ding, seconds: 0.7, decay: false, gain: 0.3)
    default: return Tone(partials: ding, seconds: 0.35, decay: false, gain: 0.25)
    }
  }

  private func wave(_ tone: Tone, length: Double? = nil) -> Data {
    let rate = 22050, frames = Int((length ?? tone.seconds) * Double(rate)), sounding = Int(tone.seconds * Double(rate))
    let scale = tone.partials.reduce(0) { $0 + $1.1 }
    var data = Data()
    func text(_ value: String) { data.append(contentsOf: value.utf8) }
    func word<T: FixedWidthInteger>(_ value: T) { var little = value.littleEndian; withUnsafeBytes(of: &little) { data.append(contentsOf: $0) } }
    text("RIFF"); word(UInt32(36 + frames * 2)); text("WAVEfmt "); word(UInt32(16))
    word(UInt16(1)); word(UInt16(1)); word(UInt32(rate)); word(UInt32(rate * 2)); word(UInt16(2)); word(UInt16(16))
    text("data"); word(UInt32(frames * 2))
    for i in 0..<frames {
      let t = Double(i) / Double(rate)
      let envelope = i >= sounding ? 0 : tone.decay ? exp(-t / (tone.seconds / 4)) : min(1, t / 0.004, (tone.seconds - t) / 0.004)
      let sample = tone.partials.reduce(0.0) { $0 + sin(t * $1.0 * 2 * .pi) * $1.1 } / scale
      word(Int16(sample * envelope * tone.gain * 32000))
    }
    return data
  }
}

File: android/app/src/main/kotlin/com/taeskim/setpad/TimingBridge.kt
package com.taeskim.setpad

import android.app.Activity
import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import android.speech.tts.TextToSpeech
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.view.WindowManager
import java.util.Locale
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import kotlin.math.PI
import kotlin.math.exp
import kotlin.math.sin

class TimingBridge(private val activity: Activity, messenger: BinaryMessenger) {
    private val channel = MethodChannel(messenger, "setpad/timing")
    private var track: AudioTrack? = null
    private var tempo: Int? = null
    private var cueTrack: AudioTrack? = null
    // Built once and replayed. Building an AudioTrack per cue put tens of
    // milliseconds between the moment a phase ends and the sound.
    private val beats = mutableMapOf<Int, AudioTrack>()
    private val cues = mutableMapOf<String, AudioTrack>()
    private var speech: TextToSpeech? = null
    private var speechReady = false
    private val handler = Handler(Looper.getMainLooper())
    private var pendingSpeech: Runnable? = null
    private var cueEndsAt = 0L
    private var spokenLocale: String? = null

    private fun prepareSpeech() {
        if (speech == null) speech = TextToSpeech(activity) { status ->
            speechReady = status == TextToSpeech.SUCCESS
        }
    }

    // Match Tempo: read half a beat after the click, capped at one second.
    // The count comes from Dart's clock and the click from this track, which
    // started a little later — near the end of the loop the click this count
    // belongs to has not sounded yet, so wait for it. A newer count replaces one
    // still waiting; if the previous word is still sounding, wait briefly for it
    // instead of dropping this count or cutting that one off.
    private fun speak(text: String, locale: String, rate: Float) {
        val engine = speech ?: return
        if (text.isEmpty() || !speechReady) return
        pendingSpeech?.let { handler.removeCallbacks(it) }; pendingSpeech = null
        val player = track
        val bpm = tempo
        var patience = 150L
        // A Tabata with no BPM still announces its rounds, so there may be no beat to wait for.
        val countRemaining = if (player != null && bpm != null) {
            val frames = (60.0 / bpm * 22050).toInt()
            val period = 60000L / bpm
            val half = minOf(30000L / bpm, 1000L)
            val at = ((player.playbackHeadPosition.toLong() and 0xffffffffL) % frames) * 1000 / 22050
            // No later than a quarter beat after the planned moment: a late word
            // never lands on the next click and lateness cannot pile up.
            patience = period / 4
            if (at > period * 3 / 4) period - at + half else maxOf(0L, half - at)
        } else 0L
        val delay = maxOf(countRemaining, cueEndsAt - SystemClock.elapsedRealtime(), 0L)
        val giveUpAt = SystemClock.elapsedRealtime() + delay + patience
        lateinit var task: Runnable
        task = Runnable {
            if (engine.isSpeaking && SystemClock.elapsedRealtime() < giveUpAt) {
                handler.postDelayed(task, 20); return@Runnable
            }
            pendingSpeech = null
            if (!engine.isSpeaking) {
                // Setting the language is a binder call; only when it changes.
                if (locale != spokenLocale) {
                    runCatching { engine.language = Locale.forLanguageTag(locale) }
                    spokenLocale = locale
                }
                engine.setSpeechRate(rate)
                engine.speak(text, TextToSpeech.QUEUE_FLUSH, null, "count")
            }
        }
        pendingSpeech = task
        handler.postDelayed(task, if (delay > 0) delay + 5 else 0)
    }

    private fun cancelSpeech() {
        pendingSpeech?.let { handler.removeCallbacks(it) }; pendingSpeech = null
        runCatching { speech?.stop() }
    }

    init {
        channel.setMethodCallHandler { call, result ->
            if (call.method == "speak") {
                runCatching {
                    speak(call.argument<String>("text") ?: "", call.argument<String>("locale") ?: "en",
                        (call.argument<Double>("rate") ?: 1.15).toFloat())
                }
                result.success(null)
            }
            else if (call.method != "configure") { result.notImplemented() }
            else try {
                val active = call.argument<Boolean>("active") == true
                // The app allows 10 BPM; anything slower than this range is a typo.
                val bpm = if (active) call.argument<Int>("bpm")?.takeIf { it in 10..300 } else null
                if (active) prepareSpeech()
                if (!active) cancelSpeech()
                if (active) activity.window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
                else activity.window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
                if (bpm != tempo) {
                    // Stopping cuts the voice. A tempo change while running (Tabata
                    // work to rest) only drops a count still waiting: the round's
                    // last number is allowed to finish.
                    if (active) dropPending() else cancelSpeech()
                    stopClick()
                    if (bpm != null) {
                        val player = beats.getOrPut(bpm) {
                            val samples = samples(click, 60.0 / bpm)
                            staticTrack(samples).also {
                                check(it.setLoopPoints(0, samples.size, -1) == AudioTrack.SUCCESS)
                            }
                        }
                        track = player
                        player.reloadStaticData()
                        // Start where the timer is inside its beat, not at a fresh
                        // click: after a pause or a re-sync the clicks stay on the grid.
                        val frames = (60.0 / bpm * 22050).toInt()
                        val at = ((call.argument<Double>("phase") ?: 0.0) * 22050).toInt()
                        if (at in 1 until frames) runCatching { player.playbackHeadPosition = at }
                        player.play(); tempo = bpm
                    }
                }
                val cue = call.argument<String>("cue")
                if (cue != null) {
                    val tone = cueTone(cue)
                    val player = cues.getOrPut(cue) { staticTrack(samples(tone)) }
                    if (cueTrack !== player) runCatching { cueTrack?.stop() }
                    cueTrack = player
                    runCatching { player.stop() }
                    player.reloadStaticData(); player.play()
                    cueEndsAt = SystemClock.elapsedRealtime() + (tone.seconds * 1000).toLong()
                    // Build the rest while this one is already sounding.
                    for (name in listOf("ready", "work", "rest", "complete")) {
                        cues.getOrPut(name) { staticTrack(samples(cueTone(name))) }
                    }
                } else if (!active) { cueTrack = null }
                result.success(null)
            } catch (_: Exception) { stop(); result.error("audioUnavailable", null, null) }
        }
    }
    // Pitch, partials, length and loudness measured from Tempo's cue recordings
    // (bpm.mp3, prebpm.mp3, end_3s.mp3), so the two apps sound like one.
    private class Tone(val partials: List<Pair<Double, Double>>, val seconds: Double, val decay: Boolean, val gain: Double)
    private val click = Tone(listOf(655.0 to 1.0, 1965.0 to 0.1), 0.12, true, 0.45)
    private val ding = listOf(1787.0 to 1.0, 2664.0 to 0.38, 1010.0 to 0.29)
    private fun cueTone(name: String) = when (name) {
        "ready" -> Tone(listOf(523.0 to 1.0, 1568.0 to 0.25), 0.08, false, 0.5)
        "work" -> Tone(listOf(1046.0 to 1.0, 3138.0 to 0.18), 0.22, false, 0.5)
        "complete" -> Tone(ding, 0.7, false, 0.3)
        else -> Tone(ding, 0.35, false, 0.25)
    }
    private fun samples(tone: Tone, length: Double = tone.seconds): ShortArray {
        val rate = 22050
        val sounding = (tone.seconds * rate).toInt()
        val scale = tone.partials.sumOf { it.second }
        return ShortArray((length * rate).toInt()) { i ->
            val t = i.toDouble() / rate
            val envelope = if (i >= sounding) 0.0 else if (tone.decay) exp(-t / (tone.seconds / 4)) else minOf(1.0, t / 0.004, (tone.seconds - t) / 0.004)
            val sample = tone.partials.sumOf { sin(t * it.first * 2 * PI) * it.second } / scale
            (sample * envelope * tone.gain * 32000).toInt().toShort()
        }
    }
    private fun staticTrack(samples: ShortArray): AudioTrack {
        val player = AudioTrack.Builder()
            .setAudioAttributes(AudioAttributes.Builder().setUsage(AudioAttributes.USAGE_MEDIA).setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION).build())
            .setAudioFormat(AudioFormat.Builder().setSampleRate(22050).setEncoding(AudioFormat.ENCODING_PCM_16BIT).setChannelMask(AudioFormat.CHANNEL_OUT_MONO).build())
            .setTransferMode(AudioTrack.MODE_STATIC).setBufferSizeInBytes(samples.size * 2).build()
        check(player.write(samples, 0, samples.size) == samples.size)
        return player
    }
    private fun dropPending() { pendingSpeech?.let { handler.removeCallbacks(it) }; pendingSpeech = null }
    private fun stopClick() { track?.let { runCatching { it.stop() } }; track = null; tempo = null }
    private fun stopBeat() { cancelSpeech(); stopClick() }
    private fun stop() { stopBeat(); runCatching { speech?.stop() }; runCatching { cueTrack?.stop() }; cueTrack = null; activity.window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON) }
    fun interrupt() { stop(); channel.invokeMethod("interrupted", null) }
    fun close() {
        stop(); speech?.shutdown(); speech = null; speechReady = false
        (beats.values + cues.values).forEach { runCatching { it.release() } }
        beats.clear(); cues.clear()
        channel.setMethodCallHandler(null)
    }
}
